src/plugins/gs-plugin-legacy-external-apps.c

```diff
@@ -0,0 +1,220 @@
+
+	/* Run plugin before the flatpak plugin because we need to setup the
+	 * external runtime version in the apps before they are actually
+	 * removed/updated */
+	gs_plugin_add_rule (plugin, GS_PLUGIN_RULE_RUN_BEFORE, "flatpak");
+}
+
+gboolean
+gs_plugin_setup (GsPlugin *plugin,
+		 GCancellable *cancellable,
+		 GError **error)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	priv->installation = flatpak_installation_new_system (cancellable,
+							      error);
+
+	if (!priv->installation)
+		return FALSE;
+
+	return TRUE;
+}
+
+void
+gs_plugin_destroy (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	g_object_unref (priv->installation);
+}
+
+static gboolean
+app_is_legacy_external_app (GsApp *app)
+{
+	guint i;
+	for (i = 0; LEGACY_EXTERNAL_APPS[i] != NULL; ++i) {
+		const char *current_id = LEGACY_EXTERNAL_APPS[i];
+		if (g_strcmp0 (gs_app_get_id (app), current_id) == 0)
+			return TRUE;
+	}
+
+	return FALSE;
+}
+
+static AsApp *
+get_installed_appstream_app (GsPlugin *plugin,
+			     GsApp *app,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	AsApp *as_app;
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	const char *deploy_dir;
+	g_autoptr(AsStore) store = NULL;
+	g_autoptr(FlatpakInstalledRef) ref = NULL;
+	g_autoptr(GFile) appstream_file = NULL;
+	g_autofree char *appstream_xml = NULL;
+	g_autofree char *appstream_path = NULL;
+
+	ref = flatpak_installation_get_installed_ref (priv->installation,
+						      FLATPAK_REF_KIND_APP,
+						      gs_app_get_flatpak_name (app),
+						      gs_app_get_flatpak_arch (app),
+						      gs_app_get_flatpak_branch (app),
+						      cancellable,
+						      error);
+
+	if (!ref)
+		return NULL;
+
+	deploy_dir = flatpak_installed_ref_get_deploy_dir (ref);
+
+	appstream_xml = g_strdup_printf ("%s.appdata.xml",
+					  gs_app_get_flatpak_name (app));
+	appstream_path = g_build_filename (deploy_dir, "files", "share",
+					   "app-info", "xmls", appstream_xml,
+					   NULL);
+	appstream_file = g_file_new_for_path (appstream_path);
+
+	store = as_store_new ();
+	as_store_set_add_flags (store,
+				AS_STORE_ADD_FLAG_USE_UNIQUE_ID |
+				AS_STORE_ADD_FLAG_USE_MERGE_HEURISTIC);
+	if (!as_store_from_file (store, appstream_file, NULL, cancellable,
+				 error))
+		return NULL;
+
+	as_app = as_store_get_app_by_id (store, gs_app_get_id (app));
+	if (!as_app)
+		g_set_error (error, AS_STORE_ERROR, AS_STORE_ERROR_FAILED,
+			     "Failed to get app %s from its own installation "
+			     "AppStream file.", gs_app_get_unique_id (app));
+
+	return g_object_ref (as_app);
+}
+
+static gboolean
+setup_ext_runtime_version (GsPlugin *plugin,
+			   GsApp *app,
+			   GCancellable *cancellable,
+			   GError **error)
+{
+	const char *runtime_version;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(AsApp) as_app = NULL;
+
+	if (!app_is_legacy_external_app (app))
+		return TRUE;
```

Comment:
That would mean that the key would/could come from the general appstream which is not sync'ed with the app's own appstream. Since we can drop it in the future the general appstream while users may still have it in the installed one, we need to use the installed one. And using the installed one means reading from disk, for all apps; to save that I filter the apps by the ones that need it.